modules/wallet/transactionbuilder.go

```diff
@@ -104,115 +104,77 @@ func calculateAmountFromOutputs(outputs []types.SiacoinOutput, fee types.Currenc
 	return amount
 }
 
-// FundSiacoinsForOutputs will add enough inputs to cover the outputs to be
-// sent in the transaction. In contrast to FundSiacoins, FundSiacoinsForOutputs
-// does not aggregate inputs into one output equaling 'amount' - with a refund,
-// potentially - for later use by an output or other transaction fee. Rather,
-// it aggregates enough inputs to cover the outputs, adds the inputs and outputs
-// to the transaction, and also generates a refund output if necessary. A miner
-// fee of 0 or greater is also taken into account in the input aggregation and
-// added to the transaction if necessary.
-func (tb *transactionBuilder) FundSiacoinsForOutputs(outputs []types.SiacoinOutput, fee types.Currency) error {
-	// dustThreshold has to be obtained separate from the lock
+func (tb *transactionBuilder) fund(amountToFund types.Currency, refundID types.SiacoinOutputID, inOutputs sortedOutputs) (types.Currency, []types.SiacoinOutputID, error) {
 	dustThreshold, err := tb.wallet.DustThreshold()
 	if err != nil {
-		return err
+		return types.Currency{}, nil, err
 	}
 
-	tb.wallet.mu.Lock()
-	defer tb.wallet.mu.Unlock()
-
 	consensusHeight, err := dbGetConsensusHeight(tb.wallet.dbTx)
 	if err != nil {
-		return err
-	}
-
-	amount := calculateAmountFromOutputs(outputs, fee)
-
-	// Add a miner fee if the passed fee was greater than 0. The fee also
-	// needs to be added to the input amount we need to aggregate.
-	if fee.Cmp64(0) > 0 {
-		tb.transaction.MinerFees = append(tb.transaction.MinerFees, fee)
+		return types.Currency{}, nil, err
 	}
 
-	so, err := tb.wallet.getSortedOutputs()
-	if err != nil {
-		return err
-	}
-
-	var fund types.Currency
 	// potentialFund tracks the balance of the wallet including outputs that
 	// have been spent in other unconfirmed transactions recently. This is to
 	// provide the user with a more useful error message in the event that they
 	// are overspending.
 	var potentialFund types.Currency
-	var spentScoids []types.SiacoinOutputID
-	for i := range so.ids {
-		scoid := so.ids[i]
-		sco := so.outputs[i]
+	// Those are needed to the caller, to spend the outputs
+	var spentInOutputs []types.SiacoinOutputID
+	// Represent the total inOutput fund
+	var fundedAmount types.Currency
+
+	for i := range inOutputs.ids {
+		inOutputID := inOutputs.ids[i]
+		inOutput := inOutputs.outputs[i]
+
 		// Check that the output can be spent.
-		if err := tb.wallet.checkOutput(tb.wallet.dbTx, consensusHeight, scoid, sco, dustThreshold); err != nil {
+		err := tb.wallet.checkOutput(tb.wallet.dbTx, consensusHeight, inOutputID, inOutput, dustThreshold)
+		if (err != nil && refundID != inOutputID) {
```

Comment:
In what situation will refundID == inOutputID? I wish when I started reading this function, I had some context about how refundID was chosen or why we need to pass it.